Core/src/org/sleuthkit/autopsy/healthmonitor/ServicesHealthMonitor.java

```diff
@@ -0,0 +1,767 @@
+            return new TimingMetric(name);
+        }
+        return null;
+    }
+    
+    /**
+     * Submit the metric that was previously obtained through getTimingMetric().
+     * Call this immediately after the section of code being timed.
+     * This method is safe to call regardless of whether the Services Health
+     * Monitor is enabled.
+     * @param metric The TimingMetric object obtained from getTimingMetric()
+     */
+    public static void submitTimingMetric(TimingMetric metric) {
+        if(isEnabled.get() && (metric != null)) {
+            metric.stopTiming();
+            try {
+                getInstance().addTimingMetric(metric);
+            } catch (HealthMonitorException ex) {
+                // We don't want calling methods to have to check for exceptions, so just log it
+                logger.log(Level.SEVERE, "Error adding timing metric", ex);
+            }
+        }
+    }
+    
+    /**
+     * Add the timing metric data to the map.
+     * @param metric The metric to add. stopTiming() should already have been called.
+     */
+    private void addTimingMetric(TimingMetric metric) throws HealthMonitorException {
+
+        // Do as little as possible within the synchronized block to minimize
+        // blocking with multiple threads.
+        synchronized(this) {
+            // There's a small check-then-act situation here where isEnabled
+            // may have changed before reaching this code. This is fine - 
+            // the map still exists and any extra data added after the monitor
+            // is disabled will be deleted if the monitor is re-enabled. This
+            // seems preferable to doing another check on isEnabled within
+            // the synchronized block.
+            if(timingInfoMap.containsKey(metric.getName())) {
+                timingInfoMap.get(metric.getName()).addMetric(metric);
+            } else {
+                timingInfoMap.put(metric.getName(), new TimingInfo(metric));
+            }
+        }
+    }
+    
+    /**
+     * Write the collected metrics to the database.
+     * @throws HealthMonitorException 
+     */
+    private void writeCurrentStateToDatabase() throws HealthMonitorException {
+                
+        Map<String, TimingInfo> timingMapCopy;
+        
+        // Do as little as possible within the synchronized block since it will
+        // block threads attempting to record metrics.
+        synchronized(this) {
+            if(! isEnabled.get()) {
+                return;
+            }
+            
+            // Make a shallow copy of the timing map. The map should be small - one entry
+            // per metric name.
+            timingMapCopy = new HashMap<>(timingInfoMap);
+            timingInfoMap.clear();
+        }
+        logger.log(Level.INFO, "Writing health monitor metrics to database");
+
+        String hostName;
+        try {
+            hostName = java.net.InetAddress.getLocalHost().getHostName();
```

Comment:
Don't know what happens under the covers when getting the host name of the local host but I've seen cases where this has to reach out to DNS servers and can be slow. Might want to consider getting this up front and storing it.